cirq/circuits/circuit.py

```diff
@@ -2200,6 +2201,31 @@ def with_noise(self, noise: 'cirq.NOISE_MODEL_LIKE') -> 'cirq.Circuit':
             c_noisy += Circuit(op_tree)
         return c_noisy
 
+    def decompose(self, add_empty_moments=True) -> Iterable['cirq.Circuit']:
+        """Decomposes circuit into a set of circuits, if it's possible (i.e. if
+        the circuit qubits can be divided into two or more groups of qubits
+        such that there are no entangling gates between them).
+
+        If this is not possible, will return the set consisting of the single
+        circuit (this one).
+        """
+        uf = networkx.utils.UnionFind(self.all_qubits())
+        for op in self.all_operations():
+            if len(op.qubits) > 1:
+                uf.union(*op.qubits)
+        qubit_sets = sorted(uf.to_sets(), key=lambda x: min(x))
+        if len(qubit_sets) == 1:
+            return (self,)
+        else:
+
+            new_moments: List[List['cirq.Moment']] = [[] for _ in qubit_sets]
+            for m in self.moments:
+                for mi, qubits in enumerate(qubit_sets):
+                    new_moment = m[qubits]
+                    if add_empty_moments or len(new_moment) > 0:
+                        new_moments[mi].append(m[qubits])
+            return (Circuit(m) for m in new_moments)
```

Comment:
We have `cirq.Circuit.zip()` Would it be worth factoring out this second part out and allow slicing on a circuit by qubit like we do for `Moment`? I guess we already use `__get_item__` for time-slicing. Maybe this can be a method like `cirq.Circuit.slice_qubits`?

Reply:
Is the use case 'getting part of the circuit which contains the given qubit'?
That means that instead of all qubit sets, only the one that contains the given qubit will be used, correct? And the output will be a single circuit.
I have added qubit iterator as a parameter to this method (and renamed it to 'factorize' because 'decompose' seems to be already taken) - would it work?

Reply:
I think the new argument clutters the factorize method a bit.

I didn't have a specific use case in mind. It just seemed like nice-to-have functionality in its own right (which is why the slicing method exists on moments) and would give feature parity between circuits and moments.

Reply:
I see. The existing version should allow this already (I don't yet see how to declutter it further though :) )

Reply:
OK, I added a separate method for slice_by_qubit and refactored the common part. Moved everything to AbstractCircuit on suggestion of maffoo, please take another look.